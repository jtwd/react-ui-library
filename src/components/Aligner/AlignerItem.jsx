import React from 'react'
import { node, number } from 'prop-types'
import styled from 'styled-components'

const StyledContainer = styled.div`
  flex: ${props => (props.flex > 0) ? props.flex : 'none'};
`

/** Flexbox layout helper for alignment */
function AlignerItem({ children, flex, ...props}) {
  return (
    <StyledContainer flex={flex} {...props}>{children}</StyledContainer>
  )
}

AlignerItem.propTypes = {
  children: node.isRequired,
  flex: number
}

AlignerItem.defaultProps = {
  flex: 'none'
}

export default AlignerItem
