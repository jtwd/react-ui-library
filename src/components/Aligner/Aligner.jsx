import React from 'react'
import { node, bool, arrayOf, number, string } from 'prop-types'
import styled from 'styled-components'
import shortid from 'shortid'

import AlignerItem from './AlignerItem'
import trimChildren from '../_theme/mixins/trimChildren'
import { xs } from "../_theme/spacers"

const StyledContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 100%;
  ${props => props.stretch && `
    align-items: stretch;
  `}
  
  ${props => props.gutter && `
    > div {
      margin: 0 ${xs};
      ${trimChildren('hor')
    }
  `}
  
`

function getAlignerContents (items, fixed, ratio, alignChildren) {
  let flexRatio = false
  // if ratio has the same amount of arguments as children then it can be used
  flexRatio = ratio && items && (ratio.length === items.length) && ratio

  return items.map((item, i) => {
    let flex = flexRatio ? flexRatio[i] : 1
    if (fixed) flex = 0

    let alignSelf = null
    // if alignChildren contains the same amount or arguments than there are children then assign the relevant one
    if (alignChildren && alignChildren.length >= items.length) alignSelf = alignChildren[i]

    return (
      <AlignerItem key={`aitem_${shortid.generate()}`} flex={`${flex}`} alignSelf={alignSelf}>{item}</AlignerItem>
    )
  })
}

/** Flexbox layout helper for alignment */
function Aligner({children, fixed, stretch, ratio, alignChildren, gutter, ...props}) {
  const contents = (children.length) ? getAlignerContents(children, fixed, ratio, alignChildren) : children
  return (
    <StyledContainer stretch={stretch} gutter={gutter} {...props}>{contents}</StyledContainer>
  )
}

Aligner.propTypes = {
  /** Content to be aligned */
  children: node.isRequired,
  /** Align individual children, 1 value for every child */
  alignChildren: arrayOf(string),
  /** Gutter between children */
  gutter: bool,
  /** No flex on multiple items */
  fixed: bool,
  /** Stretch children to fill container */
  stretch: bool,
  /** Ratio of children widths */
  ratio: arrayOf(number)
}

Aligner.defaultProps = {
  fixed: false,
  alignChildren: null,
  gutter: false,
  stretch: false,
  ratio: null
}

export default Aligner
