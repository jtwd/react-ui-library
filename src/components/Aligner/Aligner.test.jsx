import React from 'react'
import {shallow} from 'enzyme'

import Aligner from './Aligner'
import AlignerItem from './AlignerItem'

describe('Aligner component', () => {
  it('renders a center aligned single child by default', () => {
    const component = shallow(<Aligner><div>Centered</div></Aligner>)
    expect(component).toMatchSnapshot()
  })

  it('does not render a AlignerItem for single children', () => {
    const component = shallow(<Aligner><div>Centered</div></Aligner>)
    expect(component.find(AlignerItem).exists()).toBe(false)
  })

  it('renders a AlignerItem on multiple children by default', () => {
    const component = shallow(
      <Aligner>
        <div>Child 1</div>
        <div>Child 2</div>
      </Aligner>
    )
    expect(component.find(AlignerItem).length).toEqual(2)
  })

  it('renders flex of 1 on multiple children by default', () => {
    const component = shallow(
      <Aligner>
        <div>Child 1</div>
        <div>Child 2</div>
      </Aligner>
    )
    expect(component.childAt(0).props().flex).toEqual(1)
    expect(component.childAt(1).props().flex).toEqual(1)
  })

  it('renders a stretch on a single child when passed the stretch prop', () => {
    const component = shallow(
      <Aligner stretch>
        <div>Child 1</div>
      </Aligner>
    )
    expect(component).toMatchSnapshot()
  })

  it('renders a stretch on a multiple children when passed the stretch prop', () => {
    const component = shallow(
      <Aligner stretch>
        <div>Child 1</div>
        <div>Child 2</div>
      </Aligner>
    )
    expect(component).toMatchSnapshot()
  })
/*
    it('renders a center aligns multiple children without flex when passed the fixed prop', () => {
      const component = shallow(
        <Aligner fixed>
          <div>Child 1</div>
          <div>Child 2</div>
        </Aligner>
      )
      expect(component).toMatchSnapshot()
    })

  it('renders a flex values on multiple children when passed a valid ratio prop', () => {
    const component = shallow(
      <Aligner ratio={[3,7]}>
        <div>Child 1</div>
        <div>Child 2</div>
      </Aligner>
    )
    expect(component).toMatchSnapshot()
  })

  it('renders no flex values on children when passed a invalid ratio prop', () => {
    const component = shallow(
      <Aligner ratio={[2,4,3,7]}>
        <div>Child 1</div>
        <div>Child 2</div>
      </Aligner>
    )
    expect(component).toMatchSnapshot()
  })

  it('renders align-self on children when passed a valid align-children prop', () => {
    const component = shallow(
      <Aligner alignChildren={['top', 'center', 'bottom']}>
        <div>Child 1</div>
        <div>Child 2</div>
        <div>Child 3</div>
      </Aligner>
    )
    expect(component).toMatchSnapshot()
  })

  it('still renders align-self on children when passed an align-children prop with more values than children', () => {
    const component = shallow(
      <Aligner alignChildren={['top', 'center', 'bottom', 'center']}>
        <div>Child 1</div>
        <div>Child 2</div>
        <div>Child 3</div>
      </Aligner>
    )
    expect(component).toMatchSnapshot()
  })

  it('still renders no align-self on children when passed an align-children prop with less values than children', () => {
    const component = shallow(
      <Aligner alignChildrenu={['top', 'center']}>
        <div>Child 1</div>
        <div>Child 2</div>
        <div>Child 3</div>
      </Aligner>
    )
    expect(component).toMatchSnapshot()
  })

  it('renders a gutter between children when passed multiple children and the gutter prop', () => {
    const component = shallow(
      <Aligner gutter>
        <div>Child 1</div>
        <div>Child 2</div>
        <div>Child 3</div>
      </Aligner>
    )
    expect(component).toMatchSnapshot()
  })
*/
})