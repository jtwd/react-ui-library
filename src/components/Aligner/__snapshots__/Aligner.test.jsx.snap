// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aligner component renders a center aligned single child by default 1`] = `
<styled.div
  gutter={null}
  stretch={false}
>
  <div>
    Centered
  </div>
</styled.div>
`;

exports[`Aligner component renders a flex values on multiple children when passed a valid ratio prop 1`] = `
<styled.div
  gutter={null}
  stretch={false}
>
  <AlignerItem
    alignSelf={null}
    flex="3"
  >
    <div>
      Child 1
    </div>
  </AlignerItem>
  <AlignerItem
    alignSelf={null}
    flex="7"
  >
    <div>
      Child 2
    </div>
  </AlignerItem>
</styled.div>
`;

exports[`Aligner component renders align-self on children when passed a valid align-children prop 1`] = `
<styled.div
  gutter={null}
  stretch={false}
>
  <AlignerItem
    alignSelf="top"
    flex="1"
  >
    <div>
      Child 1
    </div>
  </AlignerItem>
  <AlignerItem
    alignSelf="center"
    flex="1"
  >
    <div>
      Child 2
    </div>
  </AlignerItem>
  <AlignerItem
    alignSelf="bottom"
    flex="1"
  >
    <div>
      Child 3
    </div>
  </AlignerItem>
</styled.div>
`;

exports[`Aligner component renders children with a large gutter 1`] = `
<styled.div
  gutter={null}
  gutters="lg"
  stretch={false}
>
  <AlignerItem
    alignSelf={null}
    flex="1"
  >
    <div>
      Child 1
    </div>
  </AlignerItem>
  <AlignerItem
    alignSelf={null}
    flex="1"
  >
    <div>
      Child 2
    </div>
  </AlignerItem>
  <AlignerItem
    alignSelf={null}
    flex="1"
  >
    <div>
      Child 3
    </div>
  </AlignerItem>
</styled.div>
`;

exports[`Aligner component renders children with a medium gutter 1`] = `
<styled.div
  gutter={null}
  gutters="md"
  stretch={false}
>
  <AlignerItem
    alignSelf={null}
    flex="1"
  >
    <div>
      Child 1
    </div>
  </AlignerItem>
  <AlignerItem
    alignSelf={null}
    flex="1"
  >
    <div>
      Child 2
    </div>
  </AlignerItem>
  <AlignerItem
    alignSelf={null}
    flex="1"
  >
    <div>
      Child 3
    </div>
  </AlignerItem>
</styled.div>
`;

exports[`Aligner component renders children with a small gutter 1`] = `
<styled.div
  gutter={null}
  gutters="sm"
  stretch={false}
>
  <AlignerItem
    alignSelf={null}
    flex="1"
  >
    <div>
      Child 1
    </div>
  </AlignerItem>
  <AlignerItem
    alignSelf={null}
    flex="1"
  >
    <div>
      Child 2
    </div>
  </AlignerItem>
  <AlignerItem
    alignSelf={null}
    flex="1"
  >
    <div>
      Child 3
    </div>
  </AlignerItem>
</styled.div>
`;

exports[`Aligner component renders no flex values on children when passed a invalid ratio prop 1`] = `
<styled.div
  gutter={null}
  stretch={false}
>
  <AlignerItem
    alignSelf={null}
    flex="1"
  >
    <div>
      Child 1
    </div>
  </AlignerItem>
  <AlignerItem
    alignSelf={null}
    flex="1"
  >
    <div>
      Child 2
    </div>
  </AlignerItem>
</styled.div>
`;

exports[`Aligner component renders stretch when passed the stretch prop 1`] = `
.c1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 100%;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div>
      Child 1
    </div>
  </div>
  <div
    className="c1"
  >
    <div>
      Child 2
    </div>
  </div>
</div>
`;

exports[`Aligner component still renders align-self on children when passed an align-children prop with more values than children 1`] = `
<styled.div
  gutter={null}
  stretch={false}
>
  <AlignerItem
    alignSelf="top"
    flex="1"
  >
    <div>
      Child 1
    </div>
  </AlignerItem>
  <AlignerItem
    alignSelf="center"
    flex="1"
  >
    <div>
      Child 2
    </div>
  </AlignerItem>
  <AlignerItem
    alignSelf="bottom"
    flex="1"
  >
    <div>
      Child 3
    </div>
  </AlignerItem>
</styled.div>
`;

exports[`Aligner component still renders no align-self on children when passed an align-children prop with less values than children 1`] = `
<styled.div
  gutter={null}
  stretch={false}
>
  <AlignerItem
    alignSelf={null}
    flex="1"
  >
    <div>
      Child 1
    </div>
  </AlignerItem>
  <AlignerItem
    alignSelf={null}
    flex="1"
  >
    <div>
      Child 2
    </div>
  </AlignerItem>
  <AlignerItem
    alignSelf={null}
    flex="1"
  >
    <div>
      Child 3
    </div>
  </AlignerItem>
</styled.div>
`;
